### PR TITLE
GPTBigCode (StarCoder, SantaCoder Support)

### PR DESCRIPTION
**Pull request by michaelfeil**
Adding support for GPTBigCode
- SantaCoder
- StarCoder
- WizardCoder
- tiny_starcoder_py
- ...

# Changes:
Refactored GPT2 to GPTBigCode

# Be aware of:
- I implemented a similar thing in https://github.com/OpenNMT/CTranslate2/pull/1245  . Thus, the mha_mqa converter is in numpy. It works, and can be easily refactored to torch. But a better idea would be to implement MQA anyhow, and then skip this mha_mqa function altogether.

# Speed:
While it is faster than HF, https://github.com/OpenNMT/CTranslate2 is still faster (my timings around 0.3s vllm vs 0.19s) :/

Here is some "small" benchmark.
```python
from vllm import LLM, SamplingParams

def test_gptbigcode_starcoder():
    from timeit import default_timer as t
    llm = LLM(model="bigcode/starcoder", gpu_memory_utilization=0.9)

    prompts = [
        "def hello_world",
        "def hello_name(myname: str) -> str:\n ",
    ]
    sampling_params = SamplingParams( max_tokens=32)
    llm.generate(["warmup"], sampling_params)
    s = t()
    out = llm.generate(prompts, sampling_params)
    e = t()
    print(e-s)
    assert out[1].outputs[0].text
    
    from hf_hub_ctranslate2 import GeneratorCT2fromHfHub
    llm = None
    llm = GeneratorCT2fromHfHub("michaelfeil/ct2fast-starcoder", device="cuda", compute_type="int8_float16")
    # load (my) prequantized model from HF.
    llm.generate(["warmup"], max_length=32,
        include_prompt_in_result=False
    )
    s = t()
    out2 = llm.generate(prompts, max_length=32,
        include_prompt_in_result=False
    )
    e = t()
    print(e-s)
    
if __name__ == "__main__":
    test_gptbigcode_starcoder()
```

### CODE REVIEW

File: vllm/model_executor/models/gpt_bigcode.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+            
+            def _expand_mqa_mha(qkv_array, n_head, head_dim):
+                """manipulates along axis=0 from MQA to MHA
+                inputs: qkv_array.shape=((n_heads + 2) * head_dim, hidden_dim)
+                    with n_heads for q, then 1 for k, 1 for 1 v, times head dim
+                return: qkv_array.shape=(3 * n_heads * head_dim, hidden_dim)
+                
+                TODO: no longer needed if vllm supports MQA.
+                """
+                import numpy as np
```
Comment (zhuohan123) [inline vllm/model_executor/models/gpt_bigcode.py:247]: Moving this import to the top of the file?

File: vllm/model_executor/models/gpt_bigcode.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+
+class GPTBigCodeModel(nn.Module):
+
+    def __init__(self, config: GPTBigCodeConfig):
+        super().__init__()
+        self.config = config
+        assert config.add_cross_attention == False
+        # assert config.scale_attn_by_inverse_layer_idx == False
+        # assert config.reorder_and_upcast_attn == False
```
Comment (zhuohan123) [inline vllm/model_executor/models/gpt_bigcode.py:151]: Delete these commented codes?

**Review by zhuohan123 — CHANGES_REQUESTED** [review]
Thanks for contributing to vLLM! We are very happy to see the performance gain compared to HF. This is a good working-around before we support MQA.

Besides the small comments in the code, my biggest ask is: since most of the code in `gpt_bigcode.py` is the same as in `vllm/model_executor/models/gpt2.py` except the weight loading part, can we inherit `GPT2LMHeadModel` and only changes the `load_weights` function to implement `GPTBigCodeForCausalLM`? This can help reduce redundant codes in the repo.

**Comment by WoosukKwon** [comment]
> Besides the small comments in the code, my biggest ask is: since most of the code in gpt_bigcode.py is the same as in vllm/model_executor/models/gpt2.py except the weight loading part, can we inherit GPT2LMHeadModel and only changes the load_weights function to implement GPTBigCodeForCausalLM? This can help reduce redundant codes in the repo.

I have a different thought on this. I think it's fine to keep this PR's implementation as we will eventually have the full model code in `gpt_bigcode.py`.

**Comment by zhuohan123** [comment]
After discussing with @WoosukKwon I agree it's OK to have a copy of the code as in the current format. Let's fix these smaller issues and we can merge the PR!

**Review by WoosukKwon — CHANGES_REQUESTED** [review]
Thanks for the awesome work! @michaelfeil Left minor comments.

**Comment by michaelfeil** [comment]
> After discussing with @WoosukKwon I agree it's OK to have a copy of the code as in the current format. Let's fix these smaller issues and we can merge the PR!

I agree, its HF transformer style. Lots of duplication, but making the codebase more "hackable"

**Comment by michaelfeil** [comment]
> Thanks for the awesome work! @michaelfeil Left minor comments.

Should be done :)

**Review by WoosukKwon — APPROVED** [review]
@michaelfeil Great! Thanks for your effort.

**Review by zhuohan123 — APPROVED** [review]
Thanks again for the awesome work!

**Pull request merged**